weeks/week02/section03.md

```diff
@@ -1,33 +1,222 @@
+
+### Example:
+
+```python
+test_one = my_dict.get("test_1", 1)
+print(test_one)
+test_one = my_dict.get("test_1", 1)
+print(test_one)
+```
+
+## 3.3 Updating values
+
+The value can be updated also by a key:
+
+### Syntax:
+
+```python
+d[key] = value
+```
+If there is no such key, it will be added to the `d`
+
+### Example:
+
+```python
+my_dict["test_one"] = 0
+print(my_dict)
+my_dict["test_1"] = 1
+print(my_dict)
+```
+
+## 3.4 Removing items
+
+To remove a pair from the dictionary, use the `del` statement with an exact key to delete:
+
+### Syntax:
+
+```python
+del d[key]
+```
+If there is no such key, it will raise `KeyError`
+
+### Example:
+
+```python
+del my_dict["test_1"]
+print(my_dict)
+del my_dict["test_1"]
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+    del my_dict["test_1"]
+KeyError: 'test_1'
+```
+
+`pop(key, default)` method remove `key` from the dictionary and return its value. If `key` is not in 
+the dictionary, it will return `defalt` if it is provided.
+
+
+### Example:
+
+```python
+my_dict.pop("test_two")
+my_dict.pop("test_two", "two")
+my_dict.pop("test_two")
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+    my_dict.pop("test_two")
+KeyError: 'test_two'
+```
+
+## 3.5 Extracting keys
+
+Performing `list(d)` on a dictionary returns a list of all the keys used in the dictionary, in insertion order:
+
+### Example:
+
+```python
+list(my_dict)
+```
+
+Use `sorted(d)` to return a sorted list of all the keys:
+
+### Example:
+
+```python
+sorted(my_dict)
+```
+
+To check if a key is in the dictionary, use `in`:
+
+### Example:
+
+```python
+'test_three' in my_dict
+```
+
+## 3.6 Looping
+
+When looping through dictionaries, the key and corresponding value can be retrieved at the same time using the `items()`
+method.
+
+### Example:
+
+```python
+for k, v in my_dict.items():
+    print(k, v)
+```
+
+The `keys()` method returns a dict_keys object which contains keys from a dictionary and the `values()` method returns 
+dict_values object with values of a dictionary. Both these objects are sequence objects:
+
+### Example:
+
+```python
+for k in my_dict.keys():
+    print(k)
+```
+
+```python
+for v in my_dict.values():
+    print(v)
+```
+
+## 3. Dictionary comprehension
+
+A compact way to process all or part of the elements in an iterable and return a dictionary with the results.
+
+In an example below we will generate a dictionary containing key n mapped to value n ** 2:
+
+### Example:
+
+```python
+results = {n: n ** 2 for n in range(10)}
+```
 
 ### Hands-On Task:
+1. [TI] How can you merge two dictionaries in Python?
+2. [TI] Is it possible to use a tuple as a key in the dictionary?
+3. Write a function that takes a dictionary and returns a list of all the keys.
+4. Created a nested dictionary.
+5. Create a shallow copy of a nested dictionary. Modify the original dictionary. How has the shallow copy changed? 
```

Comment:
It is not relevant to Dictionary but if we mention deep and shallow copy - imho we need to mention it in material (or even more - provide example of it)

Reply:
@SSN-SPB I am not sure if we need to copy the material from the section 1. I have added a ref note to make it clear. Is it ok?

Reply:
Agreed. week02/weeks/week02/section01.md It has this information.